integration/tag: Use subtests, make parallel and move to client unit test

vvoland opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
integration/tag: Use subtests and make parallel

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on integration/image/tag_test.go:
```diff
@@ -27,43 +27,55 @@ func TestTagUnprefixedRepoByNameOrName(t *testing.T) {
 }
 
-// ensure we don't allow the use of invalid repository names or tags; these tag operations should fail
-// TODO (yongtang): Migrate to unit tests
```
This TODO may actually be quite relevant it looks like this test is testing the client, and wouldn't even need a daemon; https://github.com/moby/moby/blob/6598cba32f4aa052ca60139197fe48404b679169/client/image_tag.go#L11-L15

vvoland replied:
Isn't the client check just a fast-path though? It should still be performed on the server side.

thaJeztah replied:
It should definitely validate on the daemon as well, but looks like these tests at least won't cover that. If we want to cover those scenarios we may need a more basic client for testing (doing a raw API call)

FWIW, we should also consider what part of this we _want_ to be validated on the client side (there's quite some cases where we're validating _too much_ on the client side, and might as well delegate things more to the API, which would (SHOULD!) validate all input.

thaJeztah replied:
Doing a quick grep, it looks like this specific error is indeed only on the client, although there's some stray validation in swarmkit (to be looked if that belongs there, or should be handled elsewhere);

```bash
git grep 'not a valid repository/tag'
client/image_tag.go:            return errors.Wrapf(err, "Error parsing reference: %q is not a valid repository/tag", source)
client/image_tag.go:            return errors.Wrapf(err, "Error parsing reference: %q is not a valid repository/tag", target)
client/image_tag_test.go:       if err == nil || err.Error() != `Error parsing reference: "aa/asdf$$^/aa" is not a valid repository/tag: invalid reference format` {
client/image_tag_test.go:       if err == nil || err.Error() != `Error parsing reference: "invalid_source_image_name_" is not a valid repository/tag: invalid reference format` {
docs/api/v1.24.md:-   **500** - error parsing reference / not a valid repository/tag: repository
integration/image/tag_test.go:          assert.Check(t, is.ErrorContains(err, "not a valid repository/tag"))
integration/image/tag_test.go:          assert.Check(t, is.ErrorContains(err, "not a valid repository/tag"))
vendor/github.com/moby/swarmkit/v2/manager/controlapi/service.go:               return status.Errorf(codes.InvalidArgument, "ContainerSpec: %q is not a valid repository/tag", image)
```

vvoland replied:
Hmmm, in this case if we were to move the check from the client side to the API then we still need it as an integration test to test the whole path.

Although it would probably be the best to actually have the unified check (same errors) on both sides, so the daemon doesn't have to deal with requests which can be determined to be invalid by the client, and still have the daemon perform the check if other client is used.

I restored the TODO and added a bit more context.

thaJeztah replied:
I did a quick test, and moving this test to the client ... makes it pass (I added a quick check to make sure it doesn't make an API call); https://github.com/vvoland/moby/pull/7

WDYT?

(yes, we should still make sure we have a test that validates the _API_ does validating as well)

vvoland replied:
LGTM; although we really should have a test for the server-side checks 🙈

thaJeztah replied:
Yes, we definitely should have one if we don't! My fear with keeping this one in integration tests is that it gives the _illusion_  that we have one, and ... we don't (at least not in this test).

@vvoland also feel free to squash my commit with yours, I think it's fine to keep them as one.

vvoland replied:
I like them separate - the first one refactors the test and the second one moves it. When squashed, it's harder to see what's actually changed and what's just moved.

thaJeztah replied:
Ah, yes, that's ok

thaJeztah committed:
integration/tag: Move to client unit test

This test was testing the client-side validation, so might as well
move it there, and validate that the client invalidates before
trying to make an API call.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah reviewed (approved):
LGTM
